futures-io/src/lib.rs

```diff
@@ -0,0 +1,305 @@
+//! Asynchronous IO
+//!
+//! This crate contains the `AsyncRead` and `AsyncWrite` traits which allow
+//! data to be read and written asynchronously.
+
+#![no_std]
+#![deny(missing_docs, missing_debug_implementations)]
+#![doc(html_root_url = "https://docs.rs/futures-io/0.2")]
+
+macro_rules! if_std {
+    ($($i:item)*) => ($(
+        #[cfg(feature = "std")]
+        $i
+    )*)
+}
+
+if_std! {
+    extern crate futures_core;
+    extern crate iovec;
+    extern crate std;
+
+    use futures_core::{Async, Poll, task};
+    use std::boxed::Box;
+    use std::io as StdIo;
+    use std::ptr;
+    use std::vec::Vec;
+
+    // Re-export IoVec for convenience
+    pub use iovec::IoVec;
+
+    // Re-export io::Error so that users don't have to deal
+    // with conflicts when `use`ing `futures::io` and `std::io`.
+    pub use StdIo::Error as Error;
+
+    /// A type used to conditionally initialize buffers passed to `AsyncRead`
+    /// methods.
+    #[derive(Debug)]
+    pub struct Initializer(bool);
+
+    impl Initializer {
+        /// Returns a new `Initializer` which will zero out buffers.
+        #[inline]
+        pub fn zeroing() -> Initializer {
+            Initializer(true)
+        }
+
+        /// Returns a new `Initializer` which will not zero out buffers.
+        ///
+        /// # Safety
+        ///
+        /// This method may only be called by `AsyncRead`ers which guarantee
+        /// that they will not read from the buffers passed to `AsyncRead`
+        /// methods, and that the return value of the method accurately reflects
+        /// the number of bytes that have been written to the head of the buffer.
+        #[inline]
+        pub unsafe fn nop() -> Initializer {
+            Initializer(false)
+        }
+
+        /// Indicates if a buffer should be initialized.
+        #[inline]
+        pub fn should_initialize(&self) -> bool {
+            self.0
+        }
+
+        /// Initializes a buffer if necessary.
+        #[inline]
+        pub fn initialize(&self, buf: &mut [u8]) {
+            if self.should_initialize() {
+                unsafe { ptr::write_bytes(buf.as_mut_ptr(), 0, buf.len()) }
+            }
+        }
+    }
+
+    /// Objects which can be read asynchronously.
+    pub trait AsyncRead {
+        /// Determines if this `AsyncRead`er can work with buffers of
+        /// uninitialized memory.
+        ///
+        /// The default implementation returns an initializer which will zero
+        /// buffers.
+        ///
+        /// # Safety
+        ///
+        /// This method is `unsafe` because and `AsyncRead`er could otherwise
+        /// return a non-zeroing `Initializer` from another `AsyncRead` type
+        /// without an `unsafe` block.
+        #[inline]
+        unsafe fn initializer(&self) -> Initializer {
```

Comment:
I thought the point of the `Initializer` having an unsafe constructor was so that this method didn't need to be `unsafe`?

Reply:
`Initializer::zeroing()` is safe, `Initializer::nop()` is `unsafe`. It is safe to implement this method using `Initializer::zeroing()`, but unsafe to implement it using `Initializer::nop()`.

This function is made `unsafe` so that users cannot implement it safely by delegating to the internally-unsafe implementation of the `initializer` function on another type.

All of this is copied from the std::io API. If you have comments, questions, or ideas about how to improve this design, please leave them [in the tracking issue](https://github.com/rust-lang/rust/issues/42788). If that API changes, we will update this one as necessary.

Reply:
Quoting from the 0.2 RFC:

> This resolves the issue of the `unsafe` on `prepare_uninitialized_buffer` being unsafe to implement, not unsafe to use (the more common, and arguably the correct meaning of `unsafe fn`).

The implementation doesn't seem to match, since even if `Initializer::zeroing()` is returned from `fn initializer()`, it doesn't matter, because it's unsafe to call `initializer()`.

I'm leaving comments here because this API is being proposed in futures, a library that I rely on, and a place where stability markers cannot protect the it. The tracking issue in std hasn't seen much traction otherwise.

Reply:
This PR does indeed address the issue I raised in the text you quoted. However, I think the working I used may have been confusing.

Currently, `unsafe fn prepare_uninitialized_buffer` is `unsafe` to implement, but not `unsafe` to call. `unsafe` on functions is intended to mean "unsafe to call", not "unsafe to implement." If an RFC like https://github.com/rust-lang/rfcs/pull/2316 were to be approved, the `prepare_uninitialized_buffer` API would be unsound.

This PR fixes this problem by moving the unsafety to the call to `Initializer::nop()`. Even if https://github.com/rust-lang/rfcs/pull/2316 were approved, this API would be sound. The `initializer` function itself is still `unsafe`, but the reason is different: it isn't that `initializer` is `unsafe` to implement, but that *calling* `initializer` is `unsafe` because it could allow you to produce values of type `Initializer::nop()` even when your `AsyncRead` implementation read from the supplied buffer.

Reply:
> Currently, unsafe fn prepare_uninitialized_buffer is unsafe to implement, but not unsafe to call.

I assume you mean that is the intent, since it definitely *is* unsafe to call the function with the current compilers.

I find the reasoning that you could get an `Initializer::nop()` back a little bit arbitrary: you don't find this sort of `unsafe` "poisoning" elsewhere. For instance, you can get a string back with `str::from_utf8_unchecked()`, but that shouldn't mean that any function that happens to wrap that is inherently unsafe.

Reply:
Consider this implementation:

```rust
struct MyUnsafePerformantRead;

impl AsyncRead for MyUnsafePerformantRead {
    fn initializer(&self) -> Initializer {
        unsafe { Initializer::nop() }
    }
    fn poll_read.... // Doesn't read from the buffer
}

struct MySafeButStillUninitRead;

impl AsyncRead for MySafeButStillUninitRead {
    fn initializer&self) -> Initializer {
        MyUnsafePerformantRead.initializer()
    }
    fn poll_read.... // Does read from the buffer (uninitialized memory), causes UB
}
```

Reply:
Yes, I imagined that is the concern, but I don't see anything that adding an `unsafe` to the function will help with. In fact, adding the `unsafe` keywords doesn't protect it at all. A wrong implementation is still wrong.

Reply:
`unsafe` prevents the unsound implementation of `MySafeButStillUninitRead`. If `MyUnsafeButPerformantRead::initializer()` is unsafe, then, `MySafeButStillUninitRead::initializer` would have to use `unsafe` in order to trigger the UB.

Reply:
I'm even more unconvinced now, since you can easily grab an `Initializer` from a completely unrelated reader, and carry it around, and safely use it via `unrelated_initializer.initialize(my_buf)` everywhere else, and trigger the UB.

As a not-arbitrary example, consider `Chain` implementation takes `Box<AsyncRead>`s, and pushes them into a vec. The naive implementation calls `self.reads[0].initializer()`, and saves it, and then uses it in later reads, instead of remembering to ask for a new one.

I think the std API gives more an illusion of protection against UB, while still quite easily allowing it to happen, even by mistake.

Reply:
> since you can easily grab an Initializer from a completely unrelated reader, and carry it around

Not without using `unsafe`. If you're making `unsafe` calls and then randomly passing around the result, in explicit violation of the invariants stated in the docs, then you're right, it is easy to cause UB.